### ISSUE TITLE
Rust can't figure out that `[Self;0]` field isn't infinitely large

### ISSUE THREAD

**Issue by rodrimati1992** [rust-lang/rust#71622]
I tried this code:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=2e98ac5ab4fca253170169b5dd7fdd27
```rust
struct Entity {
    child: [Entity;0],
}
```

I expected to see the code compile

Instead, it failed to compile with this error
```
error[E0072]: recursive type `Entity` has infinite size
 --> src/lib.rs:1:1
  |
1 | struct Entity {
  | ^^^^^^^^^^^^^ recursive type has infinite size
2 |     child: [Entity;0],
  |     ----------------- recursive without indirection
  |
  = help: insert indirection (e.g., a `Box`, `Rc`, or `&`) at some point to make `Entity` representable
```

### Meta

Tried this code in Rust 1.45.0-nightly (2020-04-26 e83f7563495dbe2629b0)

The erroring code is a reduction from code that uses const generics:
```rust
#![feature(const_generics)]
struct Entity<const L: usize> {
    foo: u8,
    child: [Entity<0>;L],
}
```
which failed with the same error:
```
error[E0072]: recursive type `Entity` has infinite size
 --> src/lib.rs:2:1
  |
2 | struct Entity<const L: usize> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ recursive type has infinite size
3 |     foo: u8,
4 |     child: [Entity<0>;L],
  |     -------------------- recursive without indirection
  |
  = help: insert indirection (e.g., a `Box`, `Rc`, or `&`) at some point to make `Entity` representable
```

**Comment by devsnek** [comment]
If this is not wontfix, I'd like to try fixing it.

**Comment by mcarton** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/11924

**Comment by rodrimati1992** [comment]
Hmm,maybe I should have created the issue with just the code that uses const generics instead (the one in the original comment).

**Comment by varkor** [comment]
This isn't specific to const generics at all: any use of a type recursively in this manner will trigger the error. I'm going to close as a duplicate, and we can reopen if this issue remains after #11924 is resolved (if indeed it's even an issue).
